policy/manager.go

```diff
@@ -121,88 +140,127 @@ func (m *Manager) monitorPolicies(ctx context.Context, evalCh chan<- *sdk.Scalin
 			}
 			continue
 
-		case policyIDs := <-m.policyIDsCh:
+		case message := <-m.policyIDsCh:
 			m.log.Trace("received policy IDs listing",
-				"num", len(policyIDs.IDs), "policy_source", policyIDs.Source)
+				"num", len(message.IDs), "policy_source", message.Source)
+
+			// Stop the handlers for policies that are no longer present.
+			m.hanldersLock.Lock()
+			maps.DeleteFunc(m.handlers, func(k PolicyID, h *handlerTracker) bool {
+				if _, ok := message.IDs[k]; !ok {
+					m.log.Trace("stopping handler for removed policy",
+						"policy_id", k, "policy_source", message.Source)
+					m.stopHandler(k)
+					return true
+				}
 
-			m.lock.Lock()
+				return false
+			})
+			m.hanldersLock.Unlock()
 
-			// Reset set of policies to keep. We will remove the policies that
-			// are not in policyIDs to reconcile our state.
-			m.keep = make(map[PolicyID]bool)
+			// Now send updates to the active policies or create new handlers
+			// for any new policies
+			for policyID, updated := range message.IDs {
 
-			// Iterate over policy IDs and create new handlers if necessary
-			for policyID := range policyIDs.IDs {
+				updatedPolicy := &sdk.ScalingPolicy{}
+				var err error
 
-				// Mark policy as must-keep so it doesn't get removed.
-				m.keep[policyID] = true
+				if !updated {
+					continue
+				}
+
+				s := m.policySources[message.Source]
+				updatedPolicy, err = s.GetLatestVersion(ctx, policyID)
+				if err != nil {
+					m.log.Error("encountered an error getting the latest version for policy", "policyID", policyID, "error", err)
+					if isUnrecoverableError(err) {
+						return err
+					}
+				}
+
+				if pht := m.handlers[policyID]; pht != nil {
```

Comment:
I think we need to hold a read lock on the `m.handlers` here in case of a concurrent `ReloadSources` call.

Reply:
The `EnforceCooldown` and `ReloadSources` methods (which I think get called from outside the monitor loop?) reads from the handlers. Mutating the map isn't atomic, so if you read it without the lock you can get weird stuff like a valid key but no value (which the runtime avoids by panicking with the message "fatal error: concurrent map read and map write").